protocols/kad/src/behaviour.rs

```diff
@@ -671,43 +768,44 @@ where
                     },
                 });
             }
+
             KademliaHandlerEvent::GetProvidersRes {
                 closer_peers,
                 provider_peers,
                 user_data,
             } => {
                 let peers = closer_peers.iter().chain(provider_peers.iter());
                 self.discovered(&user_data, &source, peers);
-                if let Some(query) = self.active_queries.get_mut(&user_data) {
-                    if let QueryInfoInner::GetProviders {
-                        pending_results, ..
-                    } = &mut query.target_mut().inner {
+                if let Some(query) = self.queries.get_mut(&user_data) {
+                    if let QueryInfo::GetProviders {
+                        providers, ..
+                    } = &mut query.inner.info {
                         for peer in provider_peers {
-                            pending_results.push(peer.node_id);
+                            providers.push(peer.node_id);
                         }
                     }
                 }
             }
+
             KademliaHandlerEvent::QueryError { user_data, .. } => {
                 // It is possible that we obtain a response for a query that has finished, which is
-                // why we may not find an entry in `self.active_queries`.
-                if let Some(query) = self.active_queries.get_mut(&user_data) {
+                // why we may not find an entry in `self.queries`.
+                if let Some(query) = self.queries.get_mut(&user_data) {
                     query.on_failure(&source)
                 }
-
-                if let Some(write) = self.active_writes.get_mut(&user_data) {
-                    write.inject_write_error(&source);
-                }
             }
+
             KademliaHandlerEvent::AddProvider { key, provider_peer } => {
-                self.queued_events.push(NetworkBehaviourAction::GenerateEvent(KademliaOut::Discovered {
-                    peer_id: provider_peer.node_id.clone(),
-                    addresses: provider_peer.multiaddrs.clone(),
-                    ty: provider_peer.connection_ty,
-                }));
+                self.queued_events.push_back(NetworkBehaviourAction::GenerateEvent(
+                    KademliaEvent::Discovered {
+                        peer_id: provider_peer.node_id.clone(),
+                        addresses: provider_peer.multiaddrs.clone(),
+                        ty: provider_peer.connection_ty,
+                    }));
+                // TODO: Expire provider records.
```

Comment:
(that seems pretty important)

Reply:
It's not new though, it was never done, I just added a TODO here. It is done as part of https://github.com/libp2p/rust-libp2p/pull/1189.